File: benchmarks/bench_sort.js
// bench_sort.js — Quicksort benchmark
// Measures: recursion, array operations, partitioning

function partition(arr, lo, hi) {
    const pivot = arr[hi];
    let i = lo;
    for (let j = lo; j < hi; j++) {
        if (arr[j] < pivot) {
            [arr[i], arr[j]] = [arr[j], arr[i]];
            i++;
        }
    }
    [arr[i], arr[hi]] = [arr[hi], arr[i]];
    return i;
}

function quicksort(arr, lo, hi) {
    if (lo < hi) {
        const p = partition(arr, lo, hi);
        quicksort(arr, lo, p - 1);
        quicksort(arr, p + 1, hi);
    }
}

const n = 500000;
const arr = new Array(n);

let seed = 42;
for (let i = 0; i < n; i++) {
    seed = (seed * 1103515245 + 12345) % 2147483648;
    arr[i] = seed % 1000000;
}

quicksort(arr, 0, n - 1);

console.log(arr[0]);
console.log(arr[n >> 1]);
console.log(arr[n - 1]);

File: benchmarks/bench_sort.rs
// bench_sort.rs — Quicksort benchmark
// Measures: recursion, array operations, partitioning

fn partition(arr: &mut [i64], lo: usize, hi: usize) -> usize {
    let pivot = arr[hi];
    let mut i = lo;
    for j in lo..hi {
        if arr[j] < pivot {
            arr.swap(i, j);
            i += 1;
        }
    }
    arr.swap(i, hi);
    i
}

fn quicksort(arr: &mut [i64], lo: usize, hi: usize) {
    if lo < hi {
        let p = partition(arr, lo, hi);
        if p > 0 {
            quicksort(arr, lo, p - 1);
        }
        quicksort(arr, p + 1, hi);
    }
}

fn main() {
    let n: usize = 500_000;
    let mut arr = vec![0i64; n];

    let mut seed: i64 = 42;
    for i in 0..n {
        seed = (seed.wrapping_mul(1103515245) + 12345) % 2147483648;
        arr[i] = seed % 1000000;
    }

    quicksort(&mut arr, 0, n - 1);

    println!("{}", arr[0]);
    println!("{}", arr[n / 2]);
    println!("{}", arr[n - 1]);
}
